assets/js/googlesitekit/datastore/user/utils.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * core/user Data store: utils.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+const INVALID_DATE_INSTANCE_ERROR = 'Date param must construct to a valid date instance or be a valid date instance itself.';
+
+/**
+ * Asserts whether a given date instance is valid or invalid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Date} date Date instance to be asserted against.
+ * @return {boolean}  True if the given date instance is valid.
+ */
+export const isValidDateInstance = ( date ) => {
+	return date instanceof Date && ! isNaN( date );
+};
+
+/**
+ * Parses the given Date instance and returns a date string (YYYY-MM-DD).
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|number|Date} date Date to parse into a string. Must be valid date value to be passed into Date constructor.
+ * @return {string}                 The parsed date string (YYYY-MM-DD).
+ */
+export const getDateString = ( date ) => {
+	const dateInstance = new Date( date );
+	invariant( isValidDateInstance( dateInstance ), INVALID_DATE_INSTANCE_ERROR );
+
+	const month = `${ dateInstance.getMonth() + 1 }`;
+	const day = `${ dateInstance.getDate() }`;
+	const year = dateInstance.getFullYear();
+
+	return [
+		year,
+		month.length < 2 ? `0${ month }` : month,
+		day.length < 2 ? `0${ day }` : day,
+	].join( '-' );
+};
+
+/**
+ * Parses the given date and returns the previous date (daysBefore).
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} relativeDate Date string (YYYY-MM-DD) to subtract days from.
+ * @param {number} daysBefore   Number of days to subtract from relativeDate.
+ * @return {string}             The date string (YYYY-MM-DD) for the previous date.
+ */
+export const getPreviousDate = ( relativeDate = '', daysBefore ) => {
+	const dateArray = relativeDate.split( '-' );
+	const isValidRelativeDate = dateArray.length === 3 && isValidDateInstance( new Date( relativeDate ) );
+	invariant( isValidRelativeDate, INVALID_DATE_INSTANCE_ERROR );
+
+	const [ year, month, day ] = dateArray;
+	const date = new Date( year, month - 1, day );
```

Comment:
Why is this splitting needed? Wouldn't `new Date( relativeDate )` be the same?

Reply:
Unfortunately the date format we use (YYYY-MM-DD) doesn't play nicely when passed directly to the date constructor, it's off by 1 day (in my locale, on Chrome). This is due to the browser interpreting the string as UTC whereas my local timezone has a negative offset from UTC, putting me in the day before. Parsing dates by string is also browser specific, explicitly passing numbers gives consistent behavior. If we want to parse as strings directly, I'd recommend pulling in a library such as `date-fns`. 

`new Date('2020-09-23')` outputs `Tue Sep 22 2020 17:00:00 GMT-0700 (Pacific Daylight Time)` which is unexpected
`new Date(2020, 8, 23)` outputs `Wed Sep 23 2020 00:00:00 GMT-0700 (Pacific Daylight Time)` which is expected

If our dates were formatted using `YYYY-M-DD` this would work as expected (at least within Chrome as it isn't detecting a UTC date):
`new Date('2020-9-23')` outputs `Wed Sep 23 2020 00:00:00 GMT-0700 (Pacific Daylight Time)`.

I could always parse the string, but figured breaking into an array was an easy way to accomplish this across environments as well as giving us quick validation (verifying that array length is 3, otherwise date is invalid).

If helpful, I can include a summarized version of this info within a comment for future use.

Reply:
Relevant discussion where this came up in a previous issue:
- https://github.com/google/site-kit-wp/issues/1394#issuecomment-623342661
- https://github.com/google/site-kit-wp/issues/1394#issuecomment-637662727

As a result we created a `parseDimensionStringToDate` that uses the separate arguments for the `Date` constructor. Unfortunately that utility is specific to Analytics which provides the date as `YYYYMMDD`.